stm32 maintenance

drowe67 opened the pull request:
Been a while since we built the stm32 code, this PR is to bring it up to date with codec2/master

1. ~~Get stm32 code building again.~~
2. Tried stm32 unittests:
```
23/23 Test #23: tst_api_demod_1600_plain_codec ...   Passed  390.79 sec

61% tests passed, 9 tests failed out of 23

Total Test time (real) = 1287.84 sec

The following tests FAILED:
	  5 - tst_ofdm_mod_plain (Failed)
	  6 - tst_ofdm_mod_ldpc (Failed)
	  7 - tst_ofdm_demod_quick (Failed)
	  8 - tst_ofdm_demod_ideal (Failed)
	  9 - tst_ofdm_demod_AWGN (Failed)
	 10 - tst_ofdm_demod_fade (Failed)
	 11 - tst_ofdm_demod_ldpc (Failed)
	 12 - tst_ofdm_demod_ldpc_AWGN (Failed)
	 13 - tst_ofdm_demod_ldpc_fade (Failed)
Errors while running CTest
```

tmiw commented:
Should we look into having the STM32 code build as part of the existing GitHub actions? Might catch issues like this sooner in the future, anyway.

tmiw commented:
@drowe67, so I just flashed this PR onto my SM1000 (rev. D) and it doesn't seem to boot. All four LEDs come on and...that's it.

Flash looks okay-ish too:

```
mooneer@ubuntu:~/freedv-gui/codec2/stm32/build_stm32$ sudo dfu-util -d 0483:df11 -c 1 -i 0 -a 0 -s 0x08000000 -D sm1000v4.bin 
dfu-util 0.9

Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
Copyright 2010-2016 Tormod Volden and Stefan Schmidt
This program is Free Software and has ABSOLUTELY NO WARRANTY
Please report bugs to http://sourceforge.net/p/dfu-util/tickets/

dfu-util: Invalid DFU suffix signature
dfu-util: A valid DFU suffix will be required in a future dfu-util release!!!
Opening DFU capable USB device...
ID 0483:df11
Run-time device DFU version 011a
Claiming USB DFU Interface...
Setting Alternate Setting #0 ...
Determining device status: state = dfuERROR, status = 10
dfuERROR, clearing status
Determining device status: state = dfuIDLE, status = 0
dfuIDLE, continuing
DFU mode device DFU version 011a
Device returned transfer size 2048
DfuSe interface name: "Internal Flash  "
Downloading to address = 0x08000000, size = 302552
Download	[=========================] 100%       302552 bytes
Download done.
File downloaded successfully
```

tmiw commented:
Also, the official v4 .bin file flashes and runs fine. The size on that one is 300,728 bytes vs. 302,552 bytes here. I didn't think 2K would make that much of a difference?

drowe67 commented:
Hi @tmiw no the extra 2k shouldn't matter much - I think we still have plenty of flash available, and the linker will bomb if we run out.  Yes I noticed quite a few warnings when I built the code for the stm32.

Thanks for giving it a try, might be something else broken in the main loop in addition to the `ofdm_` unit tests that are failing above.

tmiw commented:
@drowe67, it turns out I actually have this STM32 discovery board here at home. It seems to compare favorably to the STM32L1 series that the SM1000 currently uses. Do you see any issues offhand that would prevent me from at least trying to run the current code on it? It might allow further progress on this PR if we can get the SM1000 booting again. 👍

drowe67 commented:
@tmiw it would be worth a try, I also have a stm32L4 discovery board here.  I'm not sure what the difference in the two processors is at the code level, e.g. are they object code compatible. 

Some debugging is also possible on the SM1000 - it has a serial output that can be used to `printf()` debug strings.

drowe67 commented:
@tmiw the booting problem may be unrelated to the failing unit tests above - these just relate to the OFDM modem (used for 700D).  As the SM1000 boots in analog mode after a reflash, it shouldn't actually be running the OFDM modem.

Another possibility (common with small memory systems) running out of memory as some structure or states is too big, which makes the entire system fall over.  Common culprits are the modem stats or freedv states, and it's usually fixed by appropriate #ifdefs.  There has been a large amount of refactoring in that area.

I note on boot it [briefly opens](https://github.com/drowe67/codec2/blob/master/stm32/src/sm1000_main.c#L342) freedv in 700D mode, to work out the worst case buffer sizes.  If something inside freedv has accidentally grown beyond the available memory that would be a likely failure point. Anyway, monitoring the stream of printf() UART outputs is a good way to see where it falls over.  I hook up an USB/serial adaptor to my SM1000 for that, I think there are instructions in the source somewhere (I can dig it up if you need it).

As per https://github.com/drowe67/freedv-gui/issues/150, it would be useful to tag the next SM1000 firmware release so the (known working) source for a release binary can easily be found.

tmiw commented:
@drowe67, I was able to figure out why it wasn't booting. Once I got a UART attached I saw the following:

*This PR*:

```
M1000 VERSION: V4                                                               
SM1000 main()... stack 0x2001ff6c (147)                                         
chunks of RAM segment > 256 bytes containing start up pattern:                  
  start: 0x200010f8  end: 0x2001fc78  bytes: 125824                             
n_speech_samples: 1280 n_modem_samples: 2560                                    
n_samples_16k: 5760 storage for 4 FIFOs: 46080 bytes                            
pccm before dac/adc open: 0x10000000                                            
pccm after dac/adc open: 0x1000b400                                             
pccm after buffers: 0x10010e60 
```

*Working SM1000v4*:

```
M1000 VERSION: V4                       
SM1000 main()... stack 0x2001ff6c (147) 
chunks of RAM segment > 256 bytes containing start up pattern:
  start: 0x20000b44  end: 0x2001fc7c  bytes: 127288
n_speech_samples: 1280 n_modem_samples: 1320
n_samples_16k: 3280 storage for 4 FIFOs: 26240 bytes
pccm before dac/adc open: 0x10000000    
pccm after dac/adc open: 0x10006680     
pccm after buffers: 0x10009a20          
drivers initialised...stack: 0x2001e568 
chunks of RAM segment > 256 bytes containing start up pattern:
  start: 0x2000baf8  end: 0x2000bd7c  bytes: 644
  start: 0x2000bee8  end: 0x2000c114  bytes: 556
  start: 0x2000ea74  end: 0x2000f274  bytes: 2048                               
  start: 0x2000ff94  end: 0x20010794  bytes: 2048                               
  start: 0x20013440  end: 0x2001dc00  bytes: 42944                              
loading preferences from flash....                                              
loading default preferences....                                                 
serial: 0                                                                       
tot_period: 0                                                                   
tot_warn_period: 15                                                             
menu_freq: 800                                                                  
menu_speed: 60                                                                  
menu_vol: 2                                                                     
op_mode: 0                                                                      
prefs_changed: 0                                                                
Analog                                                                          
entering main loop...
```

`n_modem_samples` calculating to 2560 was the key. I ended up seeing the following code in ofdm.c:

```
    if (ofdm->data_mode)
        // in burst data modes we skip ahead one frame to jump over preamble
        ofdm->max_samplesperframe = 2*ofdm->samplesperframe;
```

Replacing the if statement with the following resulted in being able to boot the SM1000 with the code in this PR:

```
    if (*ofdm->data_mode != 0)
        // in burst data modes we skip ahead one frame to jump over preamble
        ofdm->max_samplesperframe = 2*ofdm->samplesperframe;
```

Hopefully this change was correct. Feel free to correct if the condition should be different. 👍

drowe67 commented:
Good catch @tmiw !  Actually looking at how that field/variable is handled elsewhere for consistency we could use:
```
if (strlen(ofdm->data_mode))
       // in burst data modes we skip ahead one frame to jump over preamble
        ofdm->max_samplesperframe = 2*ofdm->samplesperframe;
```
I probably could have coded that better - maybe trying to do too many things with one variable :thinking:

tmiw commented:
> Good catch @tmiw ! Actually looking at how that field/variable is handled elsewhere for consistency we could use:
> 
> ```
> if (strlen(ofdm->data_mode))
>        // in burst data modes we skip ahead one frame to jump over preamble
>         ofdm->max_samplesperframe = 2*ofdm->samplesperframe;
> ```
> 
> I probably could have coded that better - maybe trying to do too many things with one variable 🤔

Good to know for future reference that strlen and the other string.h functions are usable on STM32. I preferred the other approach mainly because I wasn't sure those were available at all, but this way may be easier to understand.

drowe67 commented:
Another thought on the unit tests above, this might be related to the OFDM modulator clipper/output BPF filter configuration.  IIRC we don't run the clipper and output BPF on the SM1000 as that BPF uses a lot of cycles.  A lot of work was done in that area last December for the 700E/PAPR work. So it might just be that the two test vectors (x86 and stm32) are now being generated with different clipper options.  This would also explain while the tests that depend on the ofdm modulator fail, the x86 and stm32 versions are getting different inputs.

I'll poke around some time and see if I can work out.

I can also create an account for you on the test machine with the Discovery board used for the stm32 unittests if you want to play with the system, just email me a public key.

tmiw commented:
@drowe67, just emailed my key over. Might be useful to check that out anyway 👍

tmiw commented:
@drowe67, FWIW, I was doing some testing last night (mainly increasing verbosity) and it looked like the same OFDM options were being used regardless of x86 or STM32:

```
ofdm->tx_centre = 1500
ofdm->rx_centre = 1500
ofdm->fs = 8000
ofdm->ts = 0.018
ofdm->rs = 55.5556
ofdm->tcp = 0.002
ofdm->inv_m = 0.00694444
ofdm->tx_nlower = 18
ofdm->rx_nlower = 18
ofdm->doc = 0.0436332
ofdm->timing_mx_thresh = 0.3
ofdm->nc = 17
ofdm->np = 1
ofdm->ns = 8
ofdm->bps = 2
ofdm->m = 144
ofdm->ncp = 16
ofdm->ftwindowwidth = 32
ofdm->bitsperframe = 238
ofdm->bitsperpacket = 238
ofdm->rowsperframe = 7
ofdm->samplespersymbol = 160
ofdm->samplesperframe = 1280
ofdm->max_samplesperframe = 1320
ofdm->nrxbuf = 4320
ofdm->ntxtbits = 4
ofdm->nuwbits = 10
ofdm->foff_est_gain = 0.1
ofdm->foff_est_hz = 0
ofdm->timing_mx = 0
ofdm->coarse_foff_est_hz = 0
ofdm->timing_norm = 21.1111
ofdm->mean_amp = 0
ofdm->clock_offset_counter = 0
ofdm->verbose = 3
ofdm->sample_point = 0
ofdm->timing_est = 0
ofdm->timing_valid = 0
ofdm->nin = 1280
ofdm->uw_errors = 0
ofdm->sync_counter = 0
ofdm->frame_count = 0
ofdm->sync_start = false
ofdm->sync_end = false
ofdm->sync_mode = autosync
ofdm->timing_en = true
ofdm->foff_est_en = true
ofdm->phase_est_en = true
ofdm->tx_bpf_en = true
ofdm->dpsk_en = false
ofdm->phase_est_bandwidth_mode = auto
```

However, the mean error per the UT was still reduced when I forced tx_bpf_en and clip_en to false in ofdm_create() for the embedded target.

drowe67 commented:
Snap!  I was following a similar path last weekend :slightly_smiling_face: some other ideas:
1. I hit a similar bug last year where one of the machines had some txt bits injected in the data, so the output waveforms differ.  Might be worth checking the input bits are the same.
1. I was going to plot both waveforms (look at the spectrum) and see if they look "sane", ie not just random numbers coming out of the stm32.

tmiw commented:
@drowe67, I'm not sure if this helps at all but I played both ref_mod_out.raw and mod.raw through freedv-gui and got the following:

<img width="1048" alt="Screen Shot 2021-08-01 at 3 11 24 AM" src="https://user-images.githubusercontent.com/449242/127767203-4d8ac78c-d088-4441-8cba-05098dae795d.png">

ref_mod_out.raw is the lower signal on the waterfall while mod.raw is the upper one.

drowe67 commented:
OK that's interesting - the top one looks like broadband junk, whereas the bottom looks pretty standard :thinking: Here's what I see when I plot with Octave in the time domain (mod.raw on the bottom):
```
ssh -X bear.lan
cd codec/octave
ctave:1> s=load_raw("../stm32/unittest/test_run/tst_ofdm_mod_plain/ref_mod_out.raw"); plot(s)
octave:2> mod_ref=load_raw("../stm32/unittest/test_run/tst_ofdm_mod_plain/ref_mod_out.raw");
octave:3> mod_stm32=load_raw("../stm32/unittest/test_run/tst_ofdm_mod_plain/mod.raw");
octave:4> subplot(211); plot(mod_ref); subplot(212); plot(mod_stm32);
```

![Screenshot from 2021-08-02 06-06-03](https://user-images.githubusercontent.com/45574645/127784706-05433943-4db0-43fe-bf1d-f40546a2cc06.png)

Just like your results `ref_mod_out` (bottom) looks like random or highly clipped +/-2^15 samples.

I wonder if that tx BPF is still in the code path?  That's not meant to be used for the stm32.  Perhaps the scaling/gains are messed up.  Ohh, looky here in `tst_ofdm_mod.c`
```
               ofdm_ldpc_interleave_tx(ofdm, &ldpc, tx_sams, tx_bits_char, txt_bits_char);

                for(i=0; i<Nsamperframe; i++) {
                    tx_scaled[i] = OFDM_PEAK * crealf(tx_sams[i]);
                }
```
The scaling from floats to int16s is now handled inside `ofdm.c` (e.g. compare ofdm_mod.c`) ....

tmiw commented:
@drowe67, I think I figured out what's going on with the OFDM tests. I'm rerunning all the tests to be sure and I'll update the PR accordingly.

tmiw commented:
Test results as follows:

```
Test project /home/mooneer/codec2/stm32/build_stm32
      Start  1: check_ram_limit
 1/23 Test  #1: check_ram_limit ..................   Passed    0.06 sec
      Start  2: tst_ldpc_enc
 2/23 Test  #2: tst_ldpc_enc .....................   Passed   20.13 sec
      Start  3: tst_ldpc_dec_ideal
 3/23 Test  #3: tst_ldpc_dec_ideal ...............   Passed   22.83 sec
      Start  4: tst_ldpc_dec_noise
 4/23 Test  #4: tst_ldpc_dec_noise ...............   Passed   32.31 sec
      Start  5: tst_ofdm_mod_plain
 5/23 Test  #5: tst_ofdm_mod_plain ...............   Passed   28.51 sec
      Start  6: tst_ofdm_mod_ldpc
 6/23 Test  #6: tst_ofdm_mod_ldpc ................   Passed   28.52 sec
      Start  7: tst_ofdm_demod_quick
 7/23 Test  #7: tst_ofdm_demod_quick .............   Passed   28.37 sec
      Start  8: tst_ofdm_demod_ideal
 8/23 Test  #8: tst_ofdm_demod_ideal .............   Passed   28.36 sec
      Start  9: tst_ofdm_demod_AWGN
 9/23 Test  #9: tst_ofdm_demod_AWGN ..............   Passed   45.03 sec
      Start 10: tst_ofdm_demod_fade
10/23 Test #10: tst_ofdm_demod_fade ..............   Passed  158.21 sec
      Start 11: tst_ofdm_demod_ldpc
11/23 Test #11: tst_ofdm_demod_ldpc ..............   Passed   23.84 sec
      Start 12: tst_ofdm_demod_ldpc_AWGN
12/23 Test #12: tst_ofdm_demod_ldpc_AWGN .........   Passed   48.79 sec
      Start 13: tst_ofdm_demod_ldpc_fade
13/23 Test #13: tst_ofdm_demod_ldpc_fade .........   Passed  124.36 sec
      Start 14: tst_codec2_enc_1300
14/23 Test #14: tst_codec2_enc_1300 ..............   Passed   58.84 sec
      Start 15: tst_codec2_enc_700C
15/23 Test #15: tst_codec2_enc_700C ..............   Passed   58.86 sec
      Start 16: tst_codec2_dec_1300
16/23 Test #16: tst_codec2_dec_1300 ..............   Passed  308.74 sec
      Start 17: tst_codec2_dec_700C
17/23 Test #17: tst_codec2_dec_700C ..............   Passed  307.72 sec
      Start 18: tst_api_mod_700D_TEST
18/23 Test #18: tst_api_mod_700D_TEST ............   Passed   42.68 sec
      Start 19: tst_api_mod_700D_CODEC
19/23 Test #19: tst_api_mod_700D_CODEC ...........   Passed   42.56 sec
      Start 20: tst_api_demod_700D_plain_test
20/23 Test #20: tst_api_demod_700D_plain_test ....   Passed  117.34 sec
      Start 21: tst_api_demod_700D_AWGN_test
21/23 Test #21: tst_api_demod_700D_AWGN_test .....   Passed  117.46 sec
      Start 22: tst_api_demod_700D_AWGN_codec
22/23 Test #22: tst_api_demod_700D_AWGN_codec ....   Passed   95.08 sec
      Start 23: tst_api_demod_1600_plain_codec
23/23 Test #23: tst_api_demod_1600_plain_codec ...   Passed  390.81 sec

100% tests passed, 0 tests failed out of 23

Total Test time (real) = 2129.46 sec
```

For reference, I had to do the following:

1. ofdm.c: adjust OFDM amp_scale default to match what it used for 700D.
2. tst_ofdm_demod_check: extract SNR3k field from output to fix Python errors.
3. tst_ofdm_demod_setup: remove invalid --fast command argument for cohpsk_ch calls.
4. tst_ofdm_mod.c: remove duplicate multiply by OFDM_PEAK operation.

drowemyriota commented on src/ofdm.c:
```diff
@@ -209,7 +209,7 @@ struct OFDM *ofdm_create(const struct OFDM_CONFIG *config) {
         ofdm->amp_est_mode = 0;
         ofdm->tx_bpf_en = true;
-        ofdm->amp_scale = 217E3;
+        ofdm->amp_scale = 245E3;
```
This needs to stay the same as master, as it's been carefully set up to minimise PAPR when 700D is used with compression.

The tests still pass, as both stm32 and x86 will use the same `amp_scale` :+1:

drowe67 commented:
Nice work @tmiw :+1:  I became confused when reviewing the `amp_scale` changes but think I have it restored it correctly now - you had it correct !  Have also fixed the fading tests, I had refactored `cohpsk_ch` so the command line options have changed.

When I logged into bear it had a loadav of 9.0 so I rebooted it.

Be interesting to see if 700D works on the stm32 :slightly_smiling_face:

tmiw commented:
@drowe67, I reflashed my SM1000 and transmitted/received through my KX3 attached to a dummy load (with my other radio being the Hermes Lite 2 attached to the MFJ magloop) and 700D seems like it encodes and decodes without issue. 👍

drowe67 commented:
@tmiw the test machine "bear" OS isn't booting .... so I'll re-install and set up for these tests again.  Could take a few days.  We can then test against the latest codec2/master, and maybe release some firmware for wider spread testing.

As a future project, getting 700E running might be straight forward, as it's very similar to 700D, actually a bit lighter in terms of CPU for the FEC.  I guess we'd need to build up some new unittests.

drowe67 commented:
Current status of the stm32 ctests as of https://github.com/drowe67/codec2/pull/190/commits/2fe5244fdd8d25e4ef6a5d683cd525660b03ef93

```
The following tests FAILED:
	 19 - tst_api_mod_700D_CODEC (Failed)
	 22 - tst_api_demod_700D_AWGN_codec (Failed)
```
Both appear to be something to do with the speech codec being part of the test.

tmiw commented:
@drowe67, I don't seem to have permissions to run the tests at the moment:

```
mooneer@bear:~/codec2/stm32/build_stm32$ ctest -R tst_ldpc_enc -V
UpdateCTestConfiguration  from :/home/mooneer/codec2/stm32/build_stm32/DartConfiguration.tcl
Parse Config file:/home/mooneer/codec2/stm32/build_stm32/DartConfiguration.tcl
UpdateCTestConfiguration  from :/home/mooneer/codec2/stm32/build_stm32/DartConfiguration.tcl
Parse Config file:/home/mooneer/codec2/stm32/build_stm32/DartConfiguration.tcl
Test project /home/mooneer/codec2/stm32/build_stm32
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 2
    Start 2: tst_ldpc_enc

2: Test command: /usr/bin/sh "-c" "./run_stm32_tst tst_ldpc_enc "
2: Test timeout computed to be: 1500
2: test full name: tst_ldpc_enc
2: Open On-Chip Debugger 0.11.0+dev-00294-g3d9534b8a-dirty (2021-08-06-10:47)
2: Licensed under GNU GPL v2
2: For bug reports, read
2: 	http://openocd.org/doc/doxygen/bugs.html
2: debug_level: 0
2: 
2: srst_only separate srst_nogate srst_open_drain connect_deassert_srst
2: 
2: Error: libusb_open() failed with LIBUSB_ERROR_ACCESS
2: Error: open failed
2: 
2: 
2: gdb_cmds:4: Error in sourced command file:
2: :3333: Connection timed out.
2: ERROR in run_stm32_prog! Exiting...
1/1 Test #2: tst_ldpc_enc .....................***Failed   13.49 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =  13.50 sec

The following tests FAILED:
	  2 - tst_ldpc_enc (Failed)
Errors while running CTest
```

drowe67 commented:
> @drowe67, I don't seem to have permissions to run the tests at the moment:

Ooops, consequences of how I initially set up the account.  Sorry about that.  Should be OK now, I added you to the `plugdev` group:
```
mooneer@bear:~/codec2/stm32/build_stm32$ ctest -R tst_ldpc_enc
Test project /home/mooneer/codec2/stm32/build_stm32
    Start 2: tst_ldpc_enc
1/1 Test #2: tst_ldpc_enc .....................   Passed   14.95 sec

100% tests passed, 0 tests failed out of 1
```

tmiw commented:
@drowe67, thanks, it's working now. 👍 

Anyway, it looks like the following changes were made to freedv_rx and freedv_tx:

```
freedv_passthrough_gain(freedv, 1.0);
```

```
freedv_set_eq(freedv, 1); /* for 700C/D/E & 800XA */
```

Could either of those be contributing to the UT failures?

drowe67 commented:
Thanks @tmiw - I added the former for https://github.com/drowe67/codec2/pull/193.  I think `freedv_passthrough_gain(freedv, 1.0);` should be optional in `freedv_rx`.  The EQ is autually on in `sm1000_main.c`, so probably should be in the test too. I'll work out a way to handle this (and not break https://github.com/drowe67/codec2/pull/193) ....

tmiw commented:
@drowe67, I was running ms-stm32-700e on my SM1000 earlier (which is based on this PR) and seemed to run okay. I wasn't sure if 700E support should go out as part of v5 or not, hence the separate PR (#206). 👍

drowe67 commented:
Fine Business @tmiw :+1: 

How about we make v5 include 700E, and hold off an official release until then?  I think we've met the aims of this PR (bring the stm32 code up to date) and all the tests are passing.  When we are ready, we can ask end users to try v5 (with the updates in this PR and 700E) on their sm1000s.

If you think that's OK, I'll merge this PR, and we can continue development in https://github.com/drowe67/codec2/pull/206

tmiw commented:
@drowe67, works for me. 👍
